tensorflow/core/framework/op_kernel.cc

```diff
@@ -1312,8 +1313,9 @@ Status FindKernelRegistration(
   *was_attr_mismatch = false;
 
   const string& label = GetKernelLabelAttr(node_attrs);
-
-  const string key = Key(node_op, device_type, label);
+  string device_type_alias =
+      DeviceFactory::DeviceAlias(DeviceTypeString(device_type));
+  const string key = Key(node_op, device_type, device_type_alias, label);
```

Comment:
It is a little weird to have both the device type and its alias in the key.  Instead IMO the logic should be: if the device has an alias then `Key(node_op, device_type_alias, label)` else `Key(node_op, device_type, label)`.

Reply:
@sanjoy  Thanks for the comment. I have two questions here:
  1. As previous comments: built-in devices have alias == device_name, so every device has alias, right? 
  2. I think device_type + device_type_alias can be a unique key, only device_type or device_type_alias is not enough, such as 
   *1)* if two plugins register two different device_type(XPU1, XPU2), but the same device_type_alias, in this case they will still get kernel registration/dispatch conflict.
   *2)* if plugin register a new device type, but set the device_type_alias  the same as first party device type, in this case, it will still get kernel registration/dispatch conflict

Reply:
@jzhoulon  Can you please take a look on the above comment from @sanjoy. Thanks!

Reply:
I guess this would only work if device alias is also unique and different from any device type.
@sanjoy is that what you had in mind?

Reply:
> 1) if two plugins register two different device_type(XPU1, XPU2), but the same device_type_alias, in this case they will still get kernel registration/dispatch conflict.

Do we actually want to allow this?  IMO this will be a bit counter-intuitive to have two devices have the same alias; at that point why even have two devices?

> 2) if plugin register a new device type, but set the device_type_alias the same as first party device type, in this case, it will still get kernel registration/dispatch conflict

Same question here, is this something we should allow?  If all the GPU kernels (say) work for a new device type, why even register a new device type in the first place?

Reply:
Thanks @sanjoy for your valuable discussion.

> if two plugins register two different device_type(XPU1, XPU2), but the same device_type_alias, in this case they will still get kernel registration/dispatch conflict.
Do we actually want to allow this? IMO this will be a bit counter-intuitive to have two devices have the same alias; at that point why even have two devices?

Pluggable device is open for all backend vendor, so possible there are several plugins developed by different backends installed in same system, in that case, different registered device type with same alias will possible happen ([Pluggable device support register different device type).](https://github.com/tensorflow/community/blob/master/rfcs/20200624-pluggable-device-for-tensorflow.md#supported-user-scenarios-of-pluggabledevice)

> if plugin register a new device type, but set the device_type_alias the same as first party device type, in this case, it will still get kernel registration/dispatch conflict
Same question here, is this something we should allow? If all the GPU kernels (say) work for a new device type, why even register a new device type in the first place?

Pluggable device allows backend to register as "GPU" device, so that any existed GPU model can run well w/o changing any user code ([motivation of pluggable RFC](https://github.com/tensorflow/community/blob/master/rfcs/20200624-pluggable-device-for-tensorflow.md#motivation) ), current "GPU" kernels only works for CUDA backend, but it doesn't mean it can run well for other registered "GPU" device (different GPU kernel needs different kernel language, like DPC++/SYCL/OpenCL).

Reply:
> Pluggable device is open for all backend vendor, so possible there are several plugins developed by different backends installed in same system, in that case, different registered device type with same alias will possible happen

Can we instead ask plugin authors to "namespace" their devices, as "INTEL_GPU" and "GOOGLE_BLAH" etc.?  This will avoid collisions, and might be better for readability.

> Pluggable device allows backend to register as "GPU" device

That isn't how I read the original point.  By "if plugin register a new device type, but set the device_type_alias the same as first party device type" I understood that "device type is, say, XPU and device_type_alias is GPU".  This would indeed be a problem with my suggestion to only key the kernels on "GPU" (since then XPU would pick up the CUDA kernels instead).  However, my point was that it is confusing to have XPU alias to GPU, why wouldn't you alias it to "INTEL_XPU" instead?

Reply:
Both device type and alias device type are dynamically register from plugin through stream executor C API, it can be GPU+INTEL_GPU (legacy GPU mode support), and it also can be XPU+Accelerator1 (new device type registered), it's plugin author's decision, pluggable device support both scenarios.

Reply:
Why not make this a hard failure? If author doesn't use a proper unique name, then failure. I don't see a reason to encourage bad decisions. If you want to call your device "THE_AWESOME_PROCESSOR" and so does 5 others, then well it should fail. If person foo wants to call their device "INTEL_GPU", then it should fail if Intel's actual GPU is linked in. This is effectively a global variable, behavior in C++ and elsewhere is to error if you redefine a global variable with the same name. I see this no differently.

Reply:
So is something like below check code you preferred? If yes, we can add this in "Pluggable mechanism support PR", that PR will include pluggable device registration and initialization, I think it's more straightforward to add this check in initialization phase, what do you think? @jpienaar @sanjoy 

```cpp
if (device_type == DEVICE_GPU && string::npos == device_type_alias.find("GPU")) { 
    LOG(ERROR) << "Pluggable Device initialization failure: Invalid GPU alias name, " 
               << "please make sure you are using meaningful GPU alias name with GPU keyword"; 
}
```

Reply:
Currently, device type is unique, the question here is whether the device_type_alias is also need to be unique. device_type_alias is more like one attribute of the device type(another is priority). In my original mind, I was thinking vendor can provide several backends for one device type, such as XPU(CPU, GPU..) so they can reuse the model for their multiple-backends. suppose one vendor published two plugins XPU-CPU plugin and XPU-GPU plugin, and another vendor published Accelerator-CPU, Accelerator-GPU plugin, CPU and GPU are the device_type_alias. 

For application developers, they only need to know whether the model is run on XPU or Accelerator(with tf.device("xpu:0")). If user installed a XPU-CPU plugin then the kernel will dispatched to XPU-CPU kernels, if user install a GPU plugin then it will dispatched to GPU kernels.

we can add a check if you think device_type_alias is same with the first-party device type is confusing. However, I think device_type + device_type_alias as the key is still needed. because if two plugin authors pick up the same device type alias with the other, and a vendor unique device type, they don't know it will get conflict. when python users installed this two plugin and find that it got hard failures, they can only uninstall the plugins and report to both plugin authors, although they have different device type(XPU, Accelerator).

Reply:
Since alias's main purpose is resolving conflicts of cuda kernel and cuda specific logic code. Can we make it simple as :

only those pluggable_devices registered with first party device type(e.g. "GPU") has alias device type and need to follow predefined name conversion(XX_GPU).
alias for those new device type will be ignored.
for example:
GPU + XX_GPU : kernel dispatch key -> XX_GPU.
XPU + XXX : kernel dispatch key -> XPU.
@sanjoy @jpienaar @annarev will it make sense for you? thanks.

Reply:
I don't think there is any meaning for a new device type to specify an alias name, so we may just disallow it. The main motivation for "alias" to GPU device type is that TF program with "GPU" device type runs on other GPU not only NV GPU.  Currently, "GPU" ops are implicitly dispatched to CUDA/CUDNN kernels, and other GPU needs a way to dispatch to their kernels provided in the plugin.  For new device type, there is no such issue.

Reply:
Agree, so the left concern are freedom "alias" name registration (readability issue) and key generation confusing (both "device type" and "alias" will be used), we are going to address them by simplify the design as below:

 “Alias” registration (Pluggable Device Mechanism PR）
1.	Only when pluggable device registered as first party device (e.g. “GPU”), “alias” is allowed, and must follow predefined name conversion(XX_GPU), otherwise, pluggable device initialization will fail.
2.	“Alias” for new device type is not allowed.

Key generation (Add device alias support PR 43611):
1.	If “alias” is not null (pluggable registered as first part device), kernel dispatch key -> “alias”
2.	If “alias” is null (pluggable registered as new device type), kernel dispatch key ->“device” type

@sanjoy , if you agree, we will move forward to make key generation change in this PR. Thanks!